c++/src/kj/memory.h

```diff
@@ -656,6 +657,275 @@ class SpaceFor {
   };
 };
 
+#define KJ_ASSERT_PIN_PTRS
+// Generates additional assertions for Pin/Ptr lifecycle consistency.
+
+template <typename T>
+class Ptr;
+
+// =======================================================================================
+// Pin<T>
+
+template <typename T>
+class Pin {
+  // Pin<T> holds T value with a promise that it won't be moved.
+  //
+  // Pin<T> can be dereferenced to obtain a raw pointer smart wrapper Ptr<T>.
+  // The benefit of using Pin<T>/Ptr<T> combination rather than T/&T comes from 
+  // KJ_ASSERT_PIN_PTRS enabled assertions. These verify that no Ptr<T> exist
+  // when Pin<T> is being destroyed.
+
+public:
+  inline ~Pin() {
+#ifdef KJ_ASSERT_PIN_PTRS  
+    KJ_IREQUIRE(ptrs == 0, "Pin with active references destroyed");
+#endif
+  }
+
+  inline T* operator->() { return get(); }
+  inline const T* operator->() const { return get(); }
+
+  inline Ptr<T> operator*() { return Ptr<T>(this); }
+
+private:
+  KJ_DISALLOW_COPY_AND_MOVE(Pin);
+  
+  inline Pin(T&& t): t(kj::mv(t)) {}
+
+  T t;
+
+#ifdef KJ_ASSERT_PIN_PTRS  
+  uint64_t ptrs = 0;
+#endif
+
+  inline T* get() { return &t; }
+  inline const T* get() const { return &t; }
+
+  template <typename X, typename... Params>
+  friend Pin<X> pin(Params&&... params);
+
+  template <typename X>
+  friend class Ptr;
+};
+
+template <typename T, typename... Params>
+inline Pin<T> pin(Params&&... params) {
+  return Pin<T>(T(kj::fwd<Params>(params)...));
+}
+
+template <typename T>
+class Pin<kj::Own<T>> {
+  // Pin<kj::Own<T>> represent fully owned (and also pinned) object on the heap.
```

Comment:
Could we perhaps just make it so all `Own<T>` can vend `Ptr<T>` directly? Since `Own<T>` already guarantees that its target doesn't move, this seems like it could avoid a lot of boilerplate.

Reply:
Agree here... if I'm reading Kenton's comment here correctly then I agree it would be easier...

```cpp
kj::Own<Obj> obj = kj::heap<Obj>();

kj::Pin<Obj> pinnedObj = kj::mv(obj);

// Moving the pin to a kj::Own<T> only works if it's not pinned
obj = kj::mv(pinnedObj);
```

Reply:
No, my suggestion is when you have `Own<T>`, you shouldn't need `Pin<T>` at all, you can directly create a `Ptr<T>` from an `Own<T>`.

Reply:
@mikea updated me on what was discussed in detail with regards to `kj::Pin<T>`/`kj::Own<T>`/`kj::Unique<T>`(or whatever it ends up being called) and I think we're on the same page now.